ipa-core/src/protocol/dp/distributions.rs

```diff
@@ -63,44 +61,142 @@ impl Distribution<f64> for RoundedBoxMuller {
         self.inner.sample(rng).round()
     }
 }
-
 impl From<BoxMuller> for RoundedBoxMuller {
     fn from(value: BoxMuller) -> Self {
         Self { inner: value }
     }
 }
 
+///  What follows is the implementation of sampling from the Truncated Double Geometric distribution which in the
+///  process samples from Double Geometric and Geometric distributions.
+
+///  Geometric Distribution
+/// Generates a sample from a geometric distribution with the given success probability.
+
+#[derive(Debug, PartialEq)]
+pub struct Geometric {
+    probability: f64,
+}
+impl Geometric {
+    /// Creates a new `Geometric` distribution with the given success probability.
+    pub fn new(probability: f64) -> Result<Self, Error> {
+        if probability < f64::MIN_POSITIVE {
+            return Err(Error::BadGeometricProb(probability));
+        }
+        Ok(Self { probability })
+    }
+    /// Generates a sample from the `Geometric` distribution.
+    pub fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u32 {
+        // Create a Bernoulli distribution with the specified success probability
+        let bernoulli = Bernoulli::new(self.probability).expect("Invalid probability");
+        // Generate Bernoulli random numbers until the first success
+        let mut attempts = 0;
+        while !bernoulli.sample(rng) {
+            attempts += 1;
+        }
+        attempts
+    }
+}
+impl Distribution<u32> for Geometric {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u32 {
+        self.sample(rng)
+    }
+}
+
+/// Double Geometric
+#[derive(Debug, PartialEq)]
+pub struct DoubleGeometric {
+    shift: u32,
+    geometric: Geometric,
+}
+impl DoubleGeometric {
+    /// Creates a new `DoubleGeometric` distribution with the given success probability and shift parameter.
+    pub fn new(s: f64, shift: u32) -> Result<Self, Error> {
+        if s < f64::MIN_POSITIVE {
+            return Err(Error::BadS(s));
+        }
+        let success_probability = 1.0 - E.powf(-1.0 / s);
+        Ok(Self {
+            shift,
+            geometric: Geometric::new(success_probability)?,
+        })
+    }
+    /// Generates a sample from the `DoubleGeometric` distribution
+    pub fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> i32 {
+        // Self::generate_double_geometric(self.s, self.shift, rng)
+        let attempts1 = self.geometric.sample(rng);
+        let attempts2 = self.geometric.sample(rng);
+        (self.shift + attempts1 - attempts2).try_into().unwrap()
+    }
+}
+impl Distribution<i32> for DoubleGeometric {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> i32 {
+        self.sample(rng)
+    }
+}
+
+/// Truncated Double Geometric distribution.
+#[derive(Debug, PartialEq)]
+pub struct TruncatedDoubleGeometric {
+    shift: u32, // is truncated to [0, 2*shift]
```

Comment:
I am not sure I see where it is truncated, but I like the idea of doing the math once and inside the constructor

Reply:
ah, this comment is a bit confusing. It means the later samples will be truncated to within [0,2 * shift]. I'll move it to the sample implementation and make more clear.